counter/src/main/java/org/infinispan/counter/impl/listener/CounterManagerNotificationManager.java

```diff
@@ -0,0 +1,229 @@
+import org.infinispan.counter.impl.entries.CounterValue;
+import org.infinispan.counter.logging.Log;
+import org.infinispan.executors.LimitedExecutor;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.annotation.TopologyChanged;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.notifications.cachelistener.event.TopologyChangedEvent;
+import org.infinispan.util.ByteString;
+import org.infinispan.util.concurrent.WithinThreadExecutor;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * It manages all the caches events and handles them. Also, it handles the user-specific {@link CounterListener}.
+ * <p>
+ * When a particular key is updated, its update is send to the counter, via {@link
+ * CounterEventGenerator#generate(CounterKey, CounterValue)}, and the result {@link CounterEvent} is used to notify the
+ * users {@link CounterListener}.
+ * <p>
+ * Also listens to topology changes in the cache to update the {@link WeakCounter} preferred keys, via {@link
+ * TopologyChangeListener#topologyChanged()}.
+ * <p>
+ * An user's {@link CounterListener} is invoked in sequence (i.e. only the next update is invoked when the previous one
+ * is handled) but it can be invoked in different thread.
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public class CounterManagerNotificationManager {
+
+   private static final Log log = LogFactory.getLog(CounterManagerNotificationManager.class, Log.class);
+   private final Map<ByteString, Holder> counters;
+   private volatile Executor userListenerExecutor = new WithinThreadExecutor();
+   private boolean listenersRegistered;
+
+   public CounterManagerNotificationManager() {
+      counters = new ConcurrentHashMap<>();
+   }
+
+   /**
+    * The executor to use where the user's {@link CounterListener} is invoked.
+    *
+    * @param asyncExecutor The {@link Executor} implementation.
+    */
+   public void useExecutor(Executor asyncExecutor) {
+      if (asyncExecutor == null) {
+         return;
+      }
+      userListenerExecutor = new LimitedExecutor("counter-listener", asyncExecutor, 1);
+   }
+
+   /**
+    * It registers a new counter created locally.
+    *
+    * @param counterName            The counter's name.
+    * @param generator              The counter's {@link CounterEvent} generator.
+    * @param topologyChangeListener The counter's listener to topology change. It can be {@code null}.
+    * @throws IllegalStateException If the counter with that name is already registered.
+    */
+   public void registerCounter(ByteString counterName, CounterEventGenerator generator,
+         TopologyChangeListener topologyChangeListener) {
+      if (counters.putIfAbsent(counterName, new Holder(generator, topologyChangeListener)) != null) {
+         throw new IllegalStateException();
```

Comment:
If this is likely to be triggered by wrong user code, maybe we should use a proper exception message. Otherwise, leave it as it is.

Reply:
If it is throw, it is our (read my) fault. When a counter instance is created, in the CounterManager, it registers in the notification manager. Since I have only one instance per counter, this register must happen only once (we don't support counters removal).
Maybe I should change to assert?

Reply:
No need.

Reply:
No need.